pkg/operator/controller/canary/service.go

```diff
@@ -12,20 +14,65 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// ensureCanaryService ensures the ingress canary service exists
+// ensureCanaryService ensures the ingress canary service exists.
 func (r *reconciler) ensureCanaryService(daemonsetRef metav1.OwnerReference) (bool, *corev1.Service, error) {
 	desired := desiredCanaryService(daemonsetRef)
 	haveService, current, err := r.currentCanaryService()
 	if err != nil {
 		return false, nil, err
 	}
-	if haveService {
-		return true, current, nil
+	switch {
+	case haveService:
+		if updated, err := r.updateCanaryService(current, desired); err != nil {
+			return true, current, err
+		} else if updated {
+			return r.currentCanaryService()
+		}
+	case !haveService:
+		if err := r.createCanaryService(desired); err != nil {
+			return false, nil, err
+		}
+		return true, desired, nil
 	}
-	if err := r.createCanaryService(desired); err != nil {
-		return false, nil, err
+	return true, current, nil
+}
+
+// updateCanaryService updates the canary service if an appropriate change is detected.
+func (r *reconciler) updateCanaryService(current, desired *corev1.Service) (bool, error) {
+	changed, updated := canaryServiceChanged(current, desired)
+	if !changed {
+		return false, nil
+	}
+
+	// Diff before updating because the client may mutate the object.
+	diff := cmp.Diff(current, updated, cmpopts.EquateEmpty())
+	if err := r.client.Update(context.TODO(), updated); err != nil {
+		return false, fmt.Errorf("failed to update canary service %s/%s: %v", updated.Namespace, updated.Name, err)
```

Comment:
```suggestion
		return false, fmt.Errorf("failed to update canary service %s/%s: %w", updated.Namespace, updated.Name, err)
```
(I suspect that `return false, err` would be good enough, but please use `%w` if you want to decorate the client error value.)